[FEAT] 로그인 기능 구현 API

zico24820-dev opened the pull request:
<!-- PR 제목 : [Commit Type] PR_내용 -->
<!-- PR 내용의 경우, 이슈 제목을 그대로 써도 되고, 이슈에 언급되지 않은 내용까지 써주세용 -->
<!-- ex) [FEAT] 회원 API 구현 -->

## ❗️ 이슈 번호
#28 

## 📝 작업 내용
- 로그인 및 로그아웃 API 구현 JWP 토큰 방식 
- Security  권한 설정 회원가입 , 로그인은 누구나 petmitAll()로 설정 
- 커스텀 클레임으로 user의 pk 설정 
- access , refresh 토큰 둘다 발급 
- login 후 정보는 header로 받을 수 있음 
 
## 💭 주의 사항
localhost:8080/join
localhost:8080/login 에서 토큰 값을 가져옵니다
localhost:8080/corses 에서 Bearer 토큰값을 넣어주면 됩니다. 

```json
{
    "eamil" : "",
    "passowrd": ""
}  
```

## 💡 리뷰 포인트
<img width="827" height="188" alt="image" src="https://github.com/user-attachments/assets/fe098e66-4758-4c80-b357-6a9845bf7870" />


<!-- PR 제목 : [Commit Type] PR_내용 -->
<!-- PR 내용의 경우, 이슈 제목을 그대로 써도 되고, 이슈에 언급되지 않은 내용까지 써주세용 -->
<!-- ex) [FEAT] 회원 API 구현 -->

zico24820-dev committed:
[feat]로그인 구현

JSeHoone commented on src/main/java/com/example/projectlxp/user/controller/UserController.java:
SecurityConfig에서는 "api/login"으로 로그인 URL을 설정해주신 것 같은데, Controller에는 해당 URL의 메서드가 보이지 않는데, 

SecurityConfig에서 FromLogin을 설정해줘서 그런 걸까요 ?

zico24820-dev replied:
네 맞습니다.

SecurityConfig에서 .formLogin() 설정을 추가하면, Spring Security가 그 URL(POST /api/login)을 자동으로 가로채서(Intercept) 로그인 과정을 대신 처리해 줍니다.

dnya0 commented on src/main/resources/application.yml:
`application.yml`에 db정보 올리지 마세요!
`application-local.yml` 파일을 생성하여 그 곳으로 내용 옮겨주세요

zico24820-dev committed:
[FEAT] 로그인 기능 구현 API

# Conflicts:
#	build.gradle
#	src/main/java/com/example/projectlxp/review/service/ReviewServiceImpl.java

JSeHoone commented on src/main/java/com/example/projectlxp/global/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,88 @@
+    // applcation-locla.yml 에서 설정 값 가져오기
+    public JwtTokenProvider(
+            @Value("${jwt.secret}") String secretKey,
+            @Value("${jwt.access-token-expiration-ms}") long accessTokenExpirationMs) {
```
시크릿 키랑 만료시간은 변하지 않을 것 같은데, 생성자의 변수로 받으신 이유가 있을까요 ?

AUTHORITIES_KEY 처럼 static final로 정의해도 될 것 같아서요 !

zico24820-dev replied:
우선 `AUTHORITIES_KEY`와 `secretKey`는 성격이 다릅니다.
`AUTHORITIES_KEY` : 절대 변하지 않는 고정된 문자열 상수(Constant) 

`secretKey` ,`expirationMs`는  `application.yml` 파일에서 읽어와야하는 `Configuration`  입니다.

만약 `static final`로 정의하면

@Value 는 static 필드에 값을 주입하지 못합니다.
`static` 필드는 Srping의 의존성 주입 전 Before에 JVM에 의해 클래스가 로드 될 때 메모리에 할당됩니다.

비유 

am 4시(JVM) : `static` 변수들이 메모리에 자리를 잡습니다. (`secretKey`는 null인 상태)
am 9시(Spring) : `Spring`컨테이너가 로드되고 applcation.yml을 읽습니다.
am 9시1분(Spring) : `Spring`이 `JwtTokenProvider` 객체를 생성합니다.
am 9시2분(Sprring):`@Value`를 보고 이 객체에 secretKey 값이 필요하구나 라며 생성자를 통해 값을 주입합니다.

`static` 필드는 이미 null로 초기화 되버린 상태이며 , `Spring`은 `static`영역을 관리하지 않습니다. 

이 방식이 생성자 주입방식이며, 전에 설명했던
1. 불변성
2. 테스트용이성 
3. JPA 최적화 의 방법입니다.

JSeHoone commented on src/main/java/com/example/projectlxp/user/dto/TokenResponseDTO.java:
```diff
@@ -0,0 +1,14 @@
+import lombok.Getter;
+
+@Getter
+public class TokenResponseDTO {
+
+    private String accesstoken;
+    private String tokenType = "Bearer"; // 출입증
+
+    public TokenResponseDTO(String accesstoken) {
+        this.accesstoken = accesstoken;
+    }
```
accessToken만 발급하는 것 같은데, refreshToken은 따로 발급하지 않으시나요 ?

zico24820-dev replied:
수정했습니다

dnya0 commented on src/main/java/com/example/projectlxp/global/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,88 @@
+        // yml의 secret 문자열을 SecretKry 객체로 변환
+        this.key = Keys.hmacShaKeyFor(secretKey.getBytes());
+        this.accessTokenExpirationMs = accessTokenExpirationMs;
+    }
+
+    // 토큰 생성 메서드 (로그인 성공 시 호출됨)
+    public String createToken(Authentication authentication) {
+        // Authentication 객체에서 권한 정보 가져오기
+        String authorities =
+                authentication.getAuthorities().stream()
+                        .map(GrantedAuthority::getAuthority)
+                        .collect(Collectors.joining(","));
+
+        long now = (new Date()).getTime();
+        Date validity = new Date(now + this.accessTokenExpirationMs); // 만료시간
+
+        return Jwts.builder()
+                .setSubject(authentication.getName()) // 로그인 할 아이디, email
+                .claim(AUTHORITIES_KEY, authorities) // 권한정보
+                .signWith(key, SignatureAlgorithm.HS512) // 비밀키로 서명
+                .setExpiration(validity) // 만료시간 설정
+                .compact();
+    }
```
커스텀 클레임으로 user의 pk를 보내주시면 안될까요?? 🥹

zico24820-dev replied:
추가했습니다

zico24820-dev committed:
[refactor] 로그인 코드 수정

커스텀 클레임으로 user의 pk 보내기

dnya0 commented on src/main/java/com/example/projectlxp/user/controller/UserController.java:
```diff
@@ -32,4 +29,12 @@ public ResponseEntity<String> join(
         return ResponseEntity.ok("회원가입 되셨습니다.");
     }
+
+    // 로그인 API
+    @PostMapping("/login")
+    public ResponseEntity<TokenResponseDTO> lgoin(
+            @Validated @RequestBody UserLoginRequestDTO requestDTO) {
+        String jwtToken = userService.login(requestDTO);
+        return ResponseEntity.ok(new TokenResponseDTO(jwtToken));
+    }
```
혹시 토큰들을 Http 헤더에 담아서 보내주실 수 있나요??

`Authorization: Bearer ${accessToken}`과 같은 형태로 넣어주시면 좋을 것 같습니다.

| key | value |
| :---: | :--- |
| Authorization | Bearer ${accessToken} |
| Authorization | Bearer aas84y8sakdbsaj122937ash28rhigj |

zico24820-dev committed:
[refactor] 로그인 구현

zico24820-dev committed:
[refactor] 오타수정 import 삭제
